akka-stream-tests/src/test/scala/akka/stream/scaladsl/AcknowledgeSinkSpec.scala

```diff
@@ -0,0 +1,99 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.scaladsl
+
+import akka.stream.ActorMaterializer
+import akka.stream.testkit.Utils._
+import akka.stream.testkit.{ AkkaSpec, _ }
+
+import scala.concurrent.duration._
+import scala.concurrent.{ Await, Future }
+import scala.util.control.NoStackTrace
+
+class AcknowledgeSinkSpec extends AkkaSpec {
+  implicit val mat = ActorMaterializer()
+  val ex = new RuntimeException with NoStackTrace
+
+  def assertFuture(value: Any, fb: Future[Option[Any]]): Unit =
+    Await.result(fb, 200.millis) should be(Some(value))
+
+  def assertComplete(fb: Future[Option[Any]]): Unit =
+    Await.result(fb, 200.millis) should be(None)
+
+  "An AcknowledgeSink" must {
+
+    "send the elements as result of future" in assertAllStagesStopped {
+      val receiveFun = Source(List(1, 2, 3)).runWith(Sink.queue(3))
+      assertFuture(1, receiveFun())
+      assertFuture(2, receiveFun())
+      assertFuture(3, receiveFun())
+      assertComplete(receiveFun())
+    }
+
+    "allow to have only one future waiting for result in each point of time" in assertAllStagesStopped {
+      val probe = TestPublisher.manualProbe[Int]()
+      val receiveFun = Source(probe).runWith(Sink.queue(3))
+      val sub = probe.expectSubscription()
+      val future = receiveFun()
+      val future2 = receiveFun()
+      an[Exception] shouldBe thrownBy { Await.result(future2, 200.millis) }
+      sub.sendNext(1)
+      Await.result(future, 200.millis).get should ===(1)
+      sub.sendComplete()
+    }
+
+    "wait for next element from upstream" in assertAllStagesStopped {
+      val probe = TestPublisher.manualProbe[Int]()
+      val receiveFun = Source(probe).runWith(Sink.queue(3))
+      val sub = probe.expectSubscription()
+      val future = receiveFun()
+      an[Exception] shouldBe thrownBy { Await.result(future, 200.millis) }
```

Comment:
This means feature will be succeed when element send from upstream but not earlier